zcash_client_sqlite/src/wallet.rs

```diff
@@ -519,28 +521,264 @@ pub(crate) fn get_balance(
+            SqliteClientError::CorruptedData(format!("Negative received note value: {}", value_raw))
+        })?;
+
+        let is_change = row.get::<_, bool>(2)?;
+
+        let max_priority_raw = row.get::<_, i64>(3)?;
+        let max_priority = parse_priority_code(max_priority_raw).ok_or_else(|| {
+            SqliteClientError::CorruptedData(format!(
+                "Priority code {} not recognized.",
+                max_priority_raw
+            ))
+        })?;
+
+        let received_height = row
+            .get::<_, Option<u32>>(4)
+            .map(|opt| opt.map(BlockHeight::from))?;
+
+        let mut spendable_value = NonNegativeAmount::ZERO;
+        let mut change_pending_confirmation = NonNegativeAmount::ZERO;
+        let mut value_pending_spendability = NonNegativeAmount::ZERO;
+
+        let is_spendable = any_spendable
+            && received_height.iter().any(|h| h <= &summary_height)
+            && max_priority <= ScanPriority::Scanned;
+
+        let is_pending_change = is_change && received_height.iter().all(|h| h > &summary_height);
+
+        if is_spendable {
+            spendable_value = value;
+        } else if is_pending_change {
+            change_pending_confirmation = value;
+        } else {
+            value_pending_spendability = value;
+        }
+
+        account_balances
+            .entry(account)
+            .and_modify(|bal| {
+                bal.sapling_balance.spendable_value = (bal.sapling_balance.spendable_value
+                    + spendable_value)
+                    .expect("Spendable value cannot overflow");
+                bal.sapling_balance.change_pending_confirmation =
+                    (bal.sapling_balance.change_pending_confirmation + change_pending_confirmation)
+                        .expect("Pending change value cannot overflow");
+                bal.sapling_balance.value_pending_spendability =
+                    (bal.sapling_balance.value_pending_spendability + value_pending_spendability)
+                        .expect("Value pending spendability cannot overflow");
+            })
+            .or_insert(AccountBalance {
+                sapling_balance: Balance {
+                    spendable_value,
+                    change_pending_confirmation,
+                    value_pending_spendability,
+                },
+                unshielded: NonNegativeAmount::ZERO,
+            });
     }
+
+    #[cfg(feature = "transparent-inputs")]
+    {
+        let mut stmt_transparent_balances = conn.prepare(
+            "SELECT u.received_by_account, SUM(u.value_zat)
+             FROM utxos u
+             LEFT OUTER JOIN transactions tx
+             ON tx.id_tx = u.spent_in_tx
+             WHERE u.height <= :max_height
+             AND tx.block IS NULL
+             GROUP BY u.received_by_account",
+        )?;
+        let mut rows = stmt_transparent_balances
+            .query(named_params![":max_height": u32::from(summary_height)])?;
```

Comment:
Re: the fix I suggested, this query does not currently appear to rely on `summary_height` being `chain_tip_height + 1` if `min_confirmations` is given as zero.